tests/MaxMind/Test/WebService/ClientTest.php

```diff
@@ -145,47 +183,92 @@ public function test400WithInvalidJson(): void
     }
 
     public function test400WithNoBody(): void
     {
         $this->expectException(\MaxMind\Exception\HttpException::class);
         $this->expectExceptionMessage('Received a 400 error for TestService with no body');
 
-        $this->withResponse(400, 'application/json', '');
+        $this->withResponseTestServer(400, 'application/json', '');
     }
 
     public function test400WithUnexpectedContentType(): void
     {
         $this->expectException(\MaxMind\Exception\HttpException::class);
         $this->expectExceptionMessage('Received a 400 error for TestService with the following body: text');
 
-        $this->withResponse(400, 'text/plain', 'text');
+        $this->withResponseTestServer(400, 'text/plain', 'text');
     }
 
     public function test400WithUnexpectedJson(): void
     {
         $this->expectException(\MaxMind\Exception\HttpException::class);
         $this->expectExceptionMessage('Error response contains JSON but it does not specify code or error keys: {"not":"expected"}');
 
-        $this->withResponse(400, 'application/json', '{"not":"expected"}');
+        $this->withResponseTestServer(400, 'application/json', '{"not":"expected"}');
     }
 
     public function test300(): void
     {
         $this->expectException(\MaxMind\Exception\HttpException::class);
         $this->expectExceptionMessage('Received an unexpected HTTP status (300) for TestService');
 
-        $this->withResponse(300, 'application/json', '');
+        $this->withResponseTestServer(300, 'application/json', '');
     }
 
     public function test500(): void
     {
         $this->expectException(\MaxMind\Exception\HttpException::class);
         $this->expectExceptionMessage('Received a server error (500) for TestService');
 
-        $this->withResponse(500, 'application/json', '');
+        $this->withResponseTestServer(500, 'application/json', '');
+    }
+
+    // Convenience method when you don't care about the request
+    // It runs the request through the test server.
+    // This version is used for when we want to test with an actual server.
+    private function withResponseTestServer(int $statusCode, string $contentType, string $body): ?array
+    {
+        // Set up the test server
+        $response = [
+            'status' => $statusCode,
+            'body' => $body,
+            'contentType' => $contentType,
+        ];
+        $this->setupResponse(json_encode($response));
+        $this->startTestServer();
+
+        return $this->runRequestTestServer(
+            'TestService',
+            '/path',
+            [],
+            10,
+            '0123456789',
+            [
+                'host' => 'localhost:8084',
+                'protocol' => 'http://',
+            ]
+        );
+    }
+
+    // runs the request through the test server
+    private function runRequestTestServer(
+        string $service,
+        string $path,
+        array $requestContent,
```

Comment:
I think we are no longer testing that the client is sending the correct request body. We are also not testing the HTTP method, the auth header, the other headers, etc. Using an HTTP server mocking library may make this all easier, but I suppose you could also do it with what you have by sending the expectations to the test server and checking there.

Reply:
Where are we testing that the request method, URL, headers and body are correct now? I am not seeing that.

Reply:
I still don't see where were are testing these things.